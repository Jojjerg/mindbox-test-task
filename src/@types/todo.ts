export interface ITodo {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export type TodoContextType = {
  todos: ITodo[];
  filter: IFilter;
  getTodos: () => void;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

export type IFilter = "All" | "Completed" | "Active";
