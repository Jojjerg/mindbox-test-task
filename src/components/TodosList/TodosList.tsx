import { ITodo } from "../../@types/todo";
import React from "react";
import Todo from "../Todo/Todo";

type ListProps = {
  todos: ITodo[],
  filter: string,
  updateTodo: (id: number) => void
}

const TodosList: React.FC<ListProps> = ({ todos, filter, updateTodo }) => {
  const noFilterTodos = 
    filter === "All" &&
    todos.map((todo: ITodo) => (
      <Todo
        key={todo.id}
        userId={todo.userId}
        id={todo.id}
        title={todo.title}
        completed={todo.completed}
        updateTodo={updateTodo}
      />
    ));

  const completedTodos =
    filter === "Completed" &&
    todos
      .filter((todo: ITodo) => todo.completed)
      .map((todo: ITodo) => (
        <Todo
          key={todo.id}
          userId={todo.userId}
          id={todo.id}
          title={todo.title}
          completed={todo.completed}
          updateTodo={updateTodo}
        />
      ));

  const activeTodos =
    filter === "Active" &&
    todos
      .filter((todo: ITodo) => !todo.completed)
      .map((todo: ITodo) => (
        <Todo
          key={todo.id}
          userId={todo.userId}
          id={todo.id}
          title={todo.title}
          completed={todo.completed}
          updateTodo={updateTodo}
        />
      ));

  return (
    <div>
      {!todos.length ? (
        <p>Список пуст</p>
      ) : (
        <>
          {noFilterTodos}

          {completedTodos}

          {activeTodos}
        </>
      )}
    </div>
  );
};

export default TodosList;
