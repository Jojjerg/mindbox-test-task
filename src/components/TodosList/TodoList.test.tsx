import { render, screen } from '@testing-library/react';

import { ITodo } from "../../@types/todo";
import React from "react";
import TodosList from "./TodosList";
import { createRenderer } from "react-test-renderer/shallow";

const renderer = createRenderer();

const todos = [
  {
    userId: 1,
    id: 1,
    title: "delectus aut autem",
    completed: false,
  },
  {
    userId: 1,
    id: 2,
    title: "quis ut nam facilis et officia qui",
    completed: false,
  },
  {
    userId: 1,
    id: 3,
    title: "fugiat veniam minus",
    completed: false,
  },
  {
    userId: 1,
    id: 4,
    title: "et porro tempora",
    completed: true,
  },
  {
    userId: 1,
    id: 5,
    title: "laboriosam mollitia et enim quasi adipisci quia provident illum",
    completed: false,
  },
  {
    userId: 1,
    id: 6,
    title: "qui ullam ratione quibusdam voluptatem quia omnis",
    completed: false,
  },
  {
    userId: 1,
    id: 7,
    title: "illo expedita consequatur quia in",
    completed: false,
  },
  {
    userId: 1,
    id: 8,
    title: "quo adipisci enim quam ut ab",
    completed: true,
  },
];

const updateTodo = (id: number) => {
  todos.filter((todo: ITodo) => {
    if (todo.id === id) {
      if (todo.completed) {
        todo.completed = false;
      } else {
        todo.completed = true;
      }
      return [...todos];
    }
  });
};

const defaultComponent = (filter: string) => <TodosList todos={todos} filter={filter} updateTodo={updateTodo} />;

describe("<TodosList />", () => {
  it("should render and match the snapshot", () => {
    renderer.render(defaultComponent("All"));
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it("should have 8 todos", () => {
    const { getAllByTestId } = render(defaultComponent("All"))
    expect(getAllByTestId('todo').length).toEqual(8)
  })

  it("should have 6 active todos", () => {
    const { getAllByTestId } = render(defaultComponent("Active"))
    expect(getAllByTestId('todo').length).toEqual(6)
  })

  it("should have 2 completed todos", () => {
    const { getAllByTestId } = render(defaultComponent("Completed"))
    expect(getAllByTestId('todo').length).toEqual(2)
  })
});
