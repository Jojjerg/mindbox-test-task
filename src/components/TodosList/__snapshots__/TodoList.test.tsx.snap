// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodosList /> should render and match the snapshot 1`] = `
<div>
  <React.Fragment>
    <Todo
      completed={false}
      id={1}
      title="delectus aut autem"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={false}
      id={2}
      title="quis ut nam facilis et officia qui"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={false}
      id={3}
      title="fugiat veniam minus"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={true}
      id={4}
      title="et porro tempora"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={false}
      id={5}
      title="laboriosam mollitia et enim quasi adipisci quia provident illum"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={false}
      id={6}
      title="qui ullam ratione quibusdam voluptatem quia omnis"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={false}
      id={7}
      title="illo expedita consequatur quia in"
      updateTodo={[Function]}
      userId={1}
    />
    <Todo
      completed={true}
      id={8}
      title="quo adipisci enim quam ut ab"
      updateTodo={[Function]}
      userId={1}
    />
  </React.Fragment>
</div>
`;
